Implementation of CI/CD in .NET application Using Shell Executor on Gitlab (#3360)
section-engineering-education/engineering-education: content/articles/implementation-of-ci-and-cd-in-.net-application-using-shell-executor-on-gitlab/index.md

### Introduction
Shell Executor is a basic executor that aids in the local development of the solution on the computer where GitLab Runner is installed. Windows Batch has been deprecated in favor of Bash and Windows PowerShell scripts. You'll need to set up GitLab Runner in Shell Executor mode and configure the paths to get started.

In this situation, the solution will be produced locally using Shell Executor on the machine where GitLab Runner is installed. If you're running GitLab on Linux, then you'll need to install all of the essential software on that same system.

> .Net applications on run on windows environment.

### prerequisites 
- Have a computer running Windows operating system.
- Download git into your machine. [Click here to donwload](https://git-scm.com/downloads.)
- Download `MSBuild` or use the visual studio building tools installed in the visual studio.

> - Use the same .exe files used to install MSBuild to install `VS Web Build` by running the command ` vs_buildtools.exe — add Microsoft.VisualStudio.Workload.WebBuildTools` in the command-line

Reviewer: ```suggestion
Use the same .exe files used to install MSBuild to install `VS Web Build`. We can achieve this by running the command ` vs_buildtools.exe — add Microsoft.VisualStudio.Workload.WebBuildTools` in the command-line.
```

- Download Nuget to reinstall our application's packages from the official website. [Click here to download](https://www.nuget.org/downloads)

### Table of contents
- [Installation of gitlab runners](#installation-of-gitlab-runners)
- [Configuring the runners](#configuring-the-runners)
- [How to create a .gitlab-ci.yml file](#how-to-create-a-.gitlab-ci.yml-file)
- [Conclusion](#conclusion)

### Installation of GitLab runners
To install and run GitLab Runner on Windows, you'll need Git and the User password. For more information, see the Git download link in the prerequisites section. 

`step-by-step guide`
1.  Make a folder on your computer first. For example `H:\Runners`

> 2. To install the GitLab runner, first, download the binaries for either the Windows 64-bit or 32-bit operating systems and place them in the folder you created. Please note that the following assumes that you've changed GitLab-runner.exe to GitLab-runner (optional).

Reviewer: ```suggestion
To install the GitLab Runner, first, download the binaries for either the Windows 64-bit or 32-bit operating systems. Then, place them in the folder you created. Please note that the following assumes that you've changed GitLab-runner.exe to GitLab-runner (optional).
```

4. Open a command prompt as an administrator.

5. Make registration for the runner. This can be achieved by the process of connecting a runner to one or more GitLab instances.
Registering multiple runners is as simple as repeating the register command on the same host system.

The following is the process of registering a runner.

- Run the command `./gitlab-runner.Exe register` to register a runner
- The URL of your GitLab instance must be entered (also known as the GitLab-ci coordinator URL).
- To register a runner, enter the token you received.
- Name and describe the runner. This value can be changed later through the GitLab user interface.
- Separate the tags linked with the runner using commas. This value can be changed later through the GitLab user interface.
- Provide the executor for the runner. Enter docker for the majority of usage cases.
- For projects that don't have a default image defined, you'll be requested to select one if you selected docker as your executor. GitLab-ci.yml.

> 6. Install GitLab Runner as a service and start it. You can either use the built-in System Account using the command `.\gitlab-runner.Exe installs` to install and `.\gitlab-runner.exe start` to start the runner (which is preferred) or a user account to execute the service. Use the commands `.\gitlab-runner.exe install --user "input your user name" --password "Input your password"` to install and 

Reviewer: ```suggestion
Install GitLab Runner as a service and start it. You can either use the built-in System Account using the command `.\gitlab-runner.exe installs` to install and `.\gitlab-runner.exe start` to start the runner (which is preferred). Or a user account to execute the service. Use the commands `.\gitlab-runner.exe install --user "Input your user name" --password "Input your password"` to install. And, `.\gitlab-runner.exe start` to start the runners.
```

`.\gitlab-runner.exe start` to start the runners.

### Configuring the runners
GitLab Runner is a build instance that can be used to perform jobs across several machines and transmit the results to GitLab. It can be installed on different users' computers, servers, and local devices. Once the runner has been installed, you can choose to make it shared or specific.
> You can use either specific or shared runners to service your jobs.

> **Shared Runners**- These runners are handy for jobs that have similar requirements across numerous projects. Instead of using several runners for multiple projects, you may handle multiple projects with a single or small number of Runners that are easy to maintain and update

Reviewer: ```suggestion
Shared Runners- These runners are handy for jobs that have similar requirements across many projects. Instead of using several runners for many projects, you may handle them with a single or small number of runners that are easy to maintain and update.
```

#### Understanding how we created the code above
While creating a GitLab-ci.yml file, one has to arrange the code in stages. Stages are a set of instructions that your code must follow to get to its final destination (Production). You can create any number of stages with any names in GitLab. For this, you need to put them under a certain stage key and specify what sequence you want them to run. Then GitLab will go through each of them one by one. If one of them fails, the others will not be able to run.

These stages include:
- build stage
- test stage
- deploy stage

**build stage***- This stage will construct the solution confirming that it is constructible. `$GITLAB_USER_LOGIN` In this variable, the user who initiated the job is identified by his or her username.

**test stage**- This stage will run the tests for our solution.

**deploy stage**-  In this stage, we put into practice the solution we built. From the example script above, the `CI_COMMIT_BRANCH` variable shows the name of the commit branch. These features are available in all branch pipelines, including the default branch pipelines. Merge request and tag pipelines are not supported.

### Conlusion

> From the article above, we have understood what GitLab runners are, their installation and configuration, and finally the creation of GitLab yml files. I would urge the reader to dig deeper into this topic and be able to build concrete yml files.

Reviewer: ```suggestion
In this article, we have understood what GitLab runners are, their installation and configuration.  Finally, we looked at the creation of GitLab yml files. I would urge the reader to dig deeper into this topic and be able to build concrete yml files.
```
